### PR TITLE
Update CircleCI images and add Address/Undefined Behavior sanitizers

### PR DESCRIPTION
**Pull request by kkaefer**
Caveats:
- We are currently not failing the build for Address/Undefined Behavior sanitizer because there are some lingering issues that we need to fix before.
- We are not using ccache or any other type of caching to make this faster.

### CODE REVIEW

**Comment by kkaefer** [comment]
Added ccache, and CircleCI caching, which generates one cache per job, per week, per branch.

- [ ] Android doesn't pick up `ccache` yet it seems.
- [ ] Android build show these errors:

      scripts/check-cxx11abi.sh: line 15: readelf: command not found
      scripts/check-cxx11abi.sh: line 15: c++filt: command not found

**Comment by brunoabinader** [comment]
Isn't ccache already enabled via https://github.com/mapbox/mbgl-ci-images/blob/e421fc204252a7908f17df050fcd8b840da96c78/linux/Dockerfile#L19 ?

**Comment by kkaefer** [comment]
The Android images aren't based off of the -linux image.

**Comment by kkaefer** [comment]
Note that the Android images already have ccache installed, and it also seems to fill up the cache, but they're not used on subsequent compiles.

File: circle.yml [hunk 1]
```diff
@@ -275,202 +311,211 @@ jobs:
...
+      - *setup-llvm-symbolizer
+      - *build-test
+      - *show-ccache-stats
       - run:
-          name: Build benchmark
-          command: make benchmark
+          name: Run tests
+          command: |
+            xvfb-run --server-args="-screen 0 1024x768x24" make run-test 2> >(tee sanitizer 1>&2)
+            # Unfortunately, Google Test eats the status code, so we'll have to check the output.
```
Comment (kkaefer) [inline circle.yml]: Google Test uses `fork` to execute the tests, and failing tests in those child processes probably return the correct status code. However, Google Test fails to interpret/forward that.

**Comment by friedbunny** [comment]
Can/should this be ported to the current iOS/Android release branch?

**Comment by kkaefer** [comment]
Let's not pull this into the release branch for now to avoid build instability. E.g. I just discovered a severe bug where building with Xcode-beta.app (version 9) will actually use the old compiler from Xcode.app (version 8).

**Pull request merged**